File: jlpt/native_ios/Sources/Domain/UseCases/StudyUseCases.swift
import Foundation

struct GetVocabularyStudyUseCase {
    let repository: StudyRepository
    func callAsFunction(level: JlptLevel, dueOnly: Bool) async -> AppResult<[StudyItem<Vocabulary>]> {
        await repository.getVocabularyProgress(level: level, dueOnly: dueOnly)
    }
}

struct GetKanjiStudyUseCase {
    let repository: StudyRepository
    func callAsFunction(level: JlptLevel, dueOnly: Bool) async -> AppResult<[StudyItem<Kanji>]> {
        await repository.getKanjiProgress(level: level, dueOnly: dueOnly)
    }
}

struct GetGrammarStudyUseCase {
    let repository: StudyRepository
    func callAsFunction(level: JlptLevel, dueOnly: Bool) async -> AppResult<[StudyItem<Grammar>]> {
        await repository.getGrammarProgress(level: level, dueOnly: dueOnly)
    }
}

struct UpdateStudyProgressUseCase {
    let repository: StudyRepository
    func callAsFunction(itemType: StudyItemType, itemId: Int64, status: StudyStatus) async -> AppResult<Void> {
        await repository.updateProgress(itemType: itemType, itemId: itemId, status: status)
    }
}

struct SubmitReviewUseCase {
    let repository: StudyRepository
    func callAsFunction(itemType: StudyItemType, itemId: Int64, quality: Int) async -> AppResult<ReviewResult> {
        await repository.submitReview(itemType: itemType, itemId: itemId, quality: quality)
    }
}

struct GetProgressSummaryUseCase {
    let repository: StudyRepository
    func callAsFunction() async -> AppResult<ProgressSummary> { await repository.getProgressSummary() }
}

File: jlpt/native_android/app/src/main/java/com/jlpt/android/domain/usecase/study/StudyUseCases.kt
package com.jlpt.android.domain.usecase.study

import com.jlpt.android.domain.model.Grammar
import com.jlpt.android.domain.model.JlptLevel
import com.jlpt.android.domain.model.Kanji
import com.jlpt.android.domain.model.ProgressSummary
import com.jlpt.android.domain.model.ReviewResult
import com.jlpt.android.domain.model.StudyItem
import com.jlpt.android.domain.model.StudyItemType
import com.jlpt.android.domain.model.StudyStatus
import com.jlpt.android.domain.model.Vocabulary
import com.jlpt.android.domain.repository.StudyRepository
import com.jlpt.android.domain.util.AppResult
import javax.inject.Inject

class GetVocabularyStudyUseCase @Inject constructor(private val repository: StudyRepository) {
    suspend operator fun invoke(level: JlptLevel, dueOnly: Boolean): AppResult<List<StudyItem<Vocabulary>>> =
        repository.getVocabularyProgress(level, dueOnly)
}

class GetKanjiStudyUseCase @Inject constructor(private val repository: StudyRepository) {
    suspend operator fun invoke(level: JlptLevel, dueOnly: Boolean): AppResult<List<StudyItem<Kanji>>> =
        repository.getKanjiProgress(level, dueOnly)
}

class GetGrammarStudyUseCase @Inject constructor(private val repository: StudyRepository) {
    suspend operator fun invoke(level: JlptLevel, dueOnly: Boolean): AppResult<List<StudyItem<Grammar>>> =
        repository.getGrammarProgress(level, dueOnly)
}

class UpdateStudyProgressUseCase @Inject constructor(private val repository: StudyRepository) {
    suspend operator fun invoke(itemType: StudyItemType, itemId: Long, status: StudyStatus): AppResult<Unit> =
        repository.updateProgress(itemType, itemId, status)
}

class SubmitReviewUseCase @Inject constructor(private val repository: StudyRepository) {
    suspend operator fun invoke(itemType: StudyItemType, itemId: Long, quality: Int): AppResult<ReviewResult> =
        repository.submitReview(itemType, itemId, quality)
}

class GetProgressSummaryUseCase @Inject constructor(private val repository: StudyRepository) {
    suspend operator fun invoke(): AppResult<ProgressSummary> = repository.getProgressSummary()
}
